// @pcd/passport-crypto/dist/cjs/src/passportCrypto.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PCDCrypto = void 0;
const libsodium_1 = require("./libsodium");
const utils = __importStar(require("./utils"));
const utils_1 = require("./utils");
/**
 * This class contains cryptographic primitives that are used by the PCD
 * application SDK and downstream packages.
 */
class PCDCrypto {
    static async newInstance() {
        const sodium = await (0, libsodium_1.getSodium)();
        return new PCDCrypto(sodium);
    }
    constructor(sodium) {
        this.sodium = sodium;
    }
    cryptoHash(str) {
        return (0, utils_1.arrayBufferToHexString)(this.sodium.crypto_hash(str));
    }
    generateRandomKey(bits = 256) {
        const bytes = bits / 8;
        const arrayBuffer = utils
            .getCrypto()
            .getRandomValues(new Uint8Array(bytes));
        return utils.arrayBufferToHexString(arrayBuffer);
    }
    /**
     * Combines generateSalt and argon2 function, returns both a random salt
     * and the resulting 32-byte encryption key.
     */
    generateSaltAndEncryptionKey(password) {
        const salt = this.generateSalt();
        const key = this.argon2(password, salt, 32);
        return { key, salt };
    }
    generateSalt(length = this.sodium.crypto_pwhash_SALTBYTES) {
        if (length < this.sodium.crypto_pwhash_SALTBYTES) {
            throw Error(`Salt must be at least ${this.sodium.crypto_pwhash_SALTBYTES}`);
        }
        const buffer = this.sodium.randombytes_buf(length);
        return utils.arrayBufferToHexString(buffer);
    }
    argon2(password, salt, length = 32) {
        const result = this.sodium.crypto_pwhash(length, utils.stringToArrayBuffer(password), utils.hexStringToArrayBuffer(salt), 
        // `oplimit` represents the maximum amount of computations to perform
        // for generating a key. `crypto_pwhash_OPSLIMIT_INTERACTIVE` is recommended
        // for intereactive, online applications such as Zupass.
        // Source: https://libsodium.gitbook.io/doc/password_hashing/default_phf
        this.sodium.crypto_pwhash_OPSLIMIT_INTERACTIVE, 
        // `memlimit` represents the maximum amount of RAM in bytes thet function
        // will use. It is required to use the same values for `opslimit` and `memlimit`.
        this.sodium.crypto_pwhash_MEMLIMIT_INTERACTIVE, this.sodium.crypto_pwhash_ALG_DEFAULT, "hex");
        return result;
    }
    xchacha20Encrypt(plaintext, nonce, key, assocData) {
        if (nonce.length !== 48) {
            throw Error("Nonce must be 24 bytes");
        }
        const arrayBuffer = this.sodium.crypto_aead_xchacha20poly1305_ietf_encrypt(plaintext, assocData || null, null, utils.hexStringToArrayBuffer(nonce), utils.hexStringToArrayBuffer(key));
        return utils.arrayBufferToBase64(arrayBuffer);
    }
    xchacha20Decrypt(ciphertext, nonce, key, assocData) {
        if (nonce.length !== 48) {
            throw Error("Nonce must be 24 bytes");
        }
        try {
            return this.sodium.crypto_aead_xchacha20poly1305_ietf_decrypt(null, utils.base64ToArrayBuffer(ciphertext), assocData || null, utils.hexStringToArrayBuffer(nonce), utils.hexStringToArrayBuffer(key), "text");
        }
        catch {
            return null;
        }
    }
    randombytesDeterministic(length, seed) {
        try {
            return this.sodium.randombytes_buf_deterministic(length, seed);
        }
        catch {
            return null;
        }
    }
}
exports.PCDCrypto = PCDCrypto;
