// @zk-kit/artifacts/dist/index.node.js
/**
 * @module @zk-kit/artifacts
 * @version 1.3.2
 * @file Utilities for downloading snark artifacts
 * @copyright Ethereum Foundation 2024
 * @license MIT
 * @see [Github]{@link https://github.com/privacy-scaling-explorations/snark-artifacts/tree/main/packages/artifacts}
*/
import { existsSync, createWriteStream } from 'node:fs';
import { mkdir } from 'node:fs/promises';
import os from 'node:os';
import { dirname } from 'node:path';

var Project;
(function (Project) {
    Project["SEMAPHORE_IDENTITY"] = "semaphore-identity";
    Project["POSEIDON"] = "poseidon";
    // RLN = 'rln',
    Project["SEMAPHORE"] = "semaphore";
})(Project || (Project = {}));
const projects = Object.values(Project);

async function maybeGetSnarkArtifacts$1(project, options = {}) {
    if (!projects.includes(project))
        throw new Error(`Project '${project}' is not supported`);
    options.version ??= 'latest';
    options.cdnUrl ??= 'https://unpkg.com';
    const BASE_URL = `${options.cdnUrl}/@zk-kit/${project}-artifacts@${options.version}`;
    const parameters = options.parameters
        ? `-${options.parameters.join('-')}`
        : '';
    return {
        wasm: `${BASE_URL}/${project}${parameters}.wasm`,
        zkey: `${BASE_URL}/${project}${parameters}.zkey`,
    };
}

async function download(url, outputPath) {
    const response = await fetch(url);
    if (!response.ok)
        throw new Error(`Failed to fetch ${url}: ${response.statusText}`);
    if (!response.body)
        throw new Error('Failed to get response body');
    const dir = dirname(outputPath);
    await mkdir(dir, { recursive: true });
    const fileStream = createWriteStream(outputPath);
    const reader = response.body.getReader();
    try {
        const pump = async () => {
            const { done, value } = await reader.read();
            if (done) {
                fileStream.end();
                return;
            }
            fileStream.write(Buffer.from(value));
            await pump();
        };
        await pump();
    }
    catch (error) {
        fileStream.close();
        throw error;
    }
}
// https://unpkg.com/@zk-kit/poseidon-artifacts@latest/poseidon.wasm -> @zk/poseidon-artifacts@latest/poseidon.wasm
const extractEndPath = (url) => url.substring(url.indexOf('@zk'));
async function maybeDownload(url) {
    const outputPath = `${os.tmpdir()}/${extractEndPath(url)}`;
    if (!existsSync(outputPath))
        await download(url, outputPath);
    return outputPath;
}
/**
 * Downloads SNARK artifacts (`wasm` and `zkey`) files if not already present in OS tmp folder.
 * @example
 * ```ts
 * {
 *   wasm: "/tmp/@zk-kit/semaphore-artifacts@latest/semaphore-3.wasm",
 *   zkey: "/tmp/@zk-kit/semaphore-artifacts@latest/semaphore-3.zkey"
 * }
 * ```
 * @returns {@link SnarkArtifacts}
 */
async function maybeGetSnarkArtifacts(...pars) {
    const { wasm: wasmUrl, zkey: zkeyUrl } = await maybeGetSnarkArtifacts$1(...pars);
    const [wasm, zkey] = await Promise.all([
        maybeDownload(wasmUrl),
        maybeDownload(zkeyUrl),
    ]);
    return {
        wasm,
        zkey,
    };
}

export { Project, maybeGetSnarkArtifacts, projects };
