// @zk-kit/artifacts/dist/index.browser.js
/**
 * @module @zk-kit/artifacts
 * @version 1.3.2
 * @file Utilities for downloading snark artifacts
 * @copyright Ethereum Foundation 2024
 * @license MIT
 * @see [Github]{@link https://github.com/privacy-scaling-explorations/snark-artifacts/tree/main/packages/artifacts}
*/
var Project;
(function (Project) {
    Project["SEMAPHORE_IDENTITY"] = "semaphore-identity";
    Project["POSEIDON"] = "poseidon";
    // RLN = 'rln',
    Project["SEMAPHORE"] = "semaphore";
})(Project || (Project = {}));
const projects = Object.values(Project);

async function maybeGetSnarkArtifacts(project, options = {}) {
    if (!projects.includes(project))
        throw new Error(`Project '${project}' is not supported`);
    options.version ??= 'latest';
    options.cdnUrl ??= 'https://unpkg.com';
    const BASE_URL = `${options.cdnUrl}/@zk-kit/${project}-artifacts@${options.version}`;
    const parameters = options.parameters
        ? `-${options.parameters.join('-')}`
        : '';
    return {
        wasm: `${BASE_URL}/${project}${parameters}.wasm`,
        zkey: `${BASE_URL}/${project}${parameters}.zkey`,
    };
}

export { Project, maybeGetSnarkArtifacts, projects };
