// @pcd/passport-interface/dist/cjs/src/api/requestDownloadAndDecryptStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.requestDownloadAndDecryptUpdatedStorage = exports.requestDownloadAndDecryptStorage = void 0;
const passport_crypto_1 = require("@pcd/passport-crypto");
const util_1 = require("@pcd/util");
const apiResult_1 = require("./apiResult");
const requestEncryptedStorage_1 = require("./requestEncryptedStorage");
/**
 * Downloads and decrypts a user's end-to-end encrypted backup of their
 * PCDs, given their encryption key. The server never learns the encryption
 * key.
 *
 * Never rejects. All information encoded in the resolved response.
 */
async function requestDownloadAndDecryptStorage(zupassServerUrl, encryptionKey) {
    const result = await requestDownloadAndDecryptUpdatedStorage(zupassServerUrl, encryptionKey, undefined);
    if (!result.success) {
        return { error: result.error, success: false };
    }
    if (!result.value.storage) {
        console.error("unexpectedly missing e2ee data");
        return {
            error: {
                name: apiResult_1.ERROR_NAME_BAD_RESPONSE,
                detailedMessage: "unexpectedly missing e2ee data"
            },
            success: false
        };
    }
    return {
        value: { storage: result.value.storage, revision: result.value.revision },
        success: true
    };
}
exports.requestDownloadAndDecryptStorage = requestDownloadAndDecryptStorage;
/**
 * Downloads and decrypts a user's end-to-end encrypted backup of their
 * PCDs, given their encryption key. The server never learns the encryption
 * key.
 *
 * The knownRevision indicates the previous version already known to the client.
 * If that is the latest version, no storage is returned.
 *
 * Never rejects. All information encoded in the resolved response.
 */
async function requestDownloadAndDecryptUpdatedStorage(zupassServerUrl, encryptionKey, knownRevision) {
    try {
        const encryptionKeyHash = await (0, passport_crypto_1.getHash)(encryptionKey);
        const storageResult = await (0, requestEncryptedStorage_1.requestEncryptedStorage)(zupassServerUrl, encryptionKeyHash, knownRevision);
        if (!storageResult.success) {
            console.error(`error loading e2ee data`, storageResult.error);
            return storageResult;
        }
        if (!storageResult.value.encryptedBlob) {
            if (!knownRevision) {
                console.error("missing e2ee data when downloading without revision");
                return {
                    error: {
                        name: apiResult_1.ERROR_NAME_BAD_RESPONSE,
                        detailedMessage: "missing e2ee data when downloading without revision"
                    },
                    success: false
                };
            }
            return {
                value: { revision: storageResult.value.revision },
                success: true
            };
        }
        const encryptedStorage = JSON.parse(storageResult.value.encryptedBlob);
        const decrypted = await (0, passport_crypto_1.passportDecrypt)(encryptedStorage, encryptionKey);
        return {
            value: {
                storage: JSON.parse(decrypted),
                revision: storageResult.value.revision
            },
            success: true
        };
    }
    catch (e) {
        console.error(`error loading e2ee data`, e);
        return {
            error: { name: apiResult_1.ERROR_NAME_UNKNOWN, detailedMessage: (0, util_1.getErrorMessage)(e) },
            success: false
        };
    }
}
exports.requestDownloadAndDecryptUpdatedStorage = requestDownloadAndDecryptUpdatedStorage;
