// @pcd/pcd-types/dist/cjs/src/pcd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRevealListArgument = exports.isToggleListArgument = exports.isPCDArgument = exports.isStringArrayArgument = exports.isObjectArgument = exports.isBooleanArgument = exports.isBigIntArgument = exports.isNumberArgument = exports.isStringArgument = exports.ArgumentTypeName = void 0;
/**
 * Fields of the object passed into {@link PCDPackage.prove} can only represent
 * one of the following types. {@link Unknown} is included to be used in a similar
 * way as {@code unknown} is used in TypeScript.
 */
var ArgumentTypeName;
(function (ArgumentTypeName) {
    ArgumentTypeName["String"] = "String";
    ArgumentTypeName["Number"] = "Number";
    ArgumentTypeName["BigInt"] = "BigInt";
    ArgumentTypeName["Boolean"] = "Boolean";
    ArgumentTypeName["Object"] = "Object";
    ArgumentTypeName["StringArray"] = "StringArray";
    ArgumentTypeName["PCD"] = "PCD";
    ArgumentTypeName["ToggleList"] = "ToggleList";
    ArgumentTypeName["Unknown"] = "Unknown";
})(ArgumentTypeName || (exports.ArgumentTypeName = ArgumentTypeName = {}));
function isStringArgument(arg) {
    return arg.argumentType === ArgumentTypeName.String;
}
exports.isStringArgument = isStringArgument;
function isNumberArgument(arg) {
    return arg.argumentType === ArgumentTypeName.Number;
}
exports.isNumberArgument = isNumberArgument;
function isBigIntArgument(arg) {
    return arg.argumentType === ArgumentTypeName.BigInt;
}
exports.isBigIntArgument = isBigIntArgument;
function isBooleanArgument(arg) {
    return arg.argumentType === ArgumentTypeName.Boolean;
}
exports.isBooleanArgument = isBooleanArgument;
function isObjectArgument(arg) {
    return arg.argumentType === ArgumentTypeName.Object;
}
exports.isObjectArgument = isObjectArgument;
function isStringArrayArgument(arg) {
    return arg.argumentType === ArgumentTypeName.StringArray;
}
exports.isStringArrayArgument = isStringArrayArgument;
function isPCDArgument(arg) {
    return arg.argumentType === ArgumentTypeName.PCD;
}
exports.isPCDArgument = isPCDArgument;
function isToggleListArgument(arg) {
    return (arg.argumentType === ArgumentTypeName.ToggleList &&
        arg.value !== undefined &&
        arg.value !== null &&
        typeof arg.value === "object" &&
        Object.values(arg.value).every((v) => typeof v === "boolean"));
}
exports.isToggleListArgument = isToggleListArgument;
function isRevealListArgument(arg) {
    return (arg.value !== undefined &&
        Object.keys(arg.value).every((k) => k.startsWith("reveal")));
}
exports.isRevealListArgument = isRevealListArgument;
