// @pcd/eddsa-frog-pcd/dist/cjs/src/EdDSAFrogPCDPackage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EdDSAFrogPCDPackage = exports.getDisplayOptions = exports.deserialize = exports.serialize = exports.verify = exports.prove = void 0;
const eddsa_pcd_1 = require("@pcd/eddsa-pcd");
const pcd_types_1 = require("@pcd/pcd-types");
const json_bigint_1 = __importDefault(require("json-bigint"));
const lodash_1 = __importDefault(require("lodash"));
const uuid_1 = require("uuid");
const EdDSAFrogPCD_1 = require("./EdDSAFrogPCD");
const utils_1 = require("./utils");
/**
 * Creates a new {@link EdDSAFrogPCD} by generating an {@link EdDSAFrogPCDProof}
 * and deriving an {@link EdDSAFrogPCDClaim} from the given {@link EdDSAFrogPCDArgs}.
 */
async function prove(args) {
    if (!args.privateKey.value) {
        throw new Error("missing private key");
    }
    if (!args.data.value) {
        throw new Error("missing data value");
    }
    const seralizedData = (0, utils_1.frogDataToBigInts)(args.data.value);
    // Creates an EdDSA PCD where the message is a serialized data
    const eddsaPCD = await eddsa_pcd_1.EdDSAPCDPackage.prove({
        message: {
            value: seralizedData.map((b) => b.toString()),
            argumentType: pcd_types_1.ArgumentTypeName.StringArray
        },
        privateKey: {
            value: args.privateKey.value,
            argumentType: pcd_types_1.ArgumentTypeName.String
        },
        id: {
            value: undefined,
            argumentType: pcd_types_1.ArgumentTypeName.String
        }
    });
    const id = args.id.value ?? (0, uuid_1.v4)();
    return new EdDSAFrogPCD_1.EdDSAFrogPCD(id, { data: args.data.value }, { eddsaPCD });
}
exports.prove = prove;
/**
 * Verifies an EdDSA Frog PCD by checking that its {@link EdDSAFrogPCDClaim} corresponds to
 * its {@link EdDSAFrogPCDProof}. If they match, the function returns true, otherwise false.
 * In most cases, verifying the validity of the PCD with this function is not enough.
 * It may also be necessary to check the public key of the
 * entity that signed the claim and verify the authenticity of the entity.
 */
async function verify(pcd) {
    const messageDerivedFromClaim = (0, utils_1.frogDataToBigInts)(pcd.claim.data);
    return (lodash_1.default.isEqual(messageDerivedFromClaim, pcd.proof.eddsaPCD.claim.message) &&
        eddsa_pcd_1.EdDSAPCDPackage.verify(pcd.proof.eddsaPCD));
}
exports.verify = verify;
/**
 * Serializes an {@link EdDSAFrogPCD}.
 * @param pcd The EdDSA Frog PCD to be serialized.
 * @returns The serialized version of the EdDSA Frog PCD.
 */
async function serialize(pcd) {
    const serializedEdDSAPCD = await eddsa_pcd_1.EdDSAPCDPackage.serialize(pcd.proof.eddsaPCD);
    return {
        type: EdDSAFrogPCD_1.EdDSAFrogPCDTypeName,
        pcd: (0, json_bigint_1.default)().stringify({
            id: pcd.id,
            eddsaPCD: serializedEdDSAPCD,
            data: pcd.claim.data
        })
    };
}
exports.serialize = serialize;
/**
 * Deserializes a serialized {@link EdDSAFrogPCD}.
 * @param serialized The serialized PCD to deserialize.
 * @returns The deserialized version of the EdDSA Frog PCD.
 */
async function deserialize(serialized) {
    const deserializedWrapper = (0, json_bigint_1.default)().parse(serialized);
    const deserializedEdDSAPCD = await eddsa_pcd_1.EdDSAPCDPackage.deserialize(deserializedWrapper.eddsaPCD.pcd);
    return new EdDSAFrogPCD_1.EdDSAFrogPCD(deserializedWrapper.id, { data: deserializedWrapper.data }, { eddsaPCD: deserializedEdDSAPCD });
}
exports.deserialize = deserialize;
/**
 * Provides the information about the {@link EdDSAFrogPCD} that will be displayed
 * to users on Zupass.
 * @param pcd The EdDSA Frog PCD instance.
 * @returns The information to be displayed, specifically `header` and `displayName`.
 */
function getDisplayOptions(pcd) {
    const frogData = (0, utils_1.getEdDSAFrogData)(pcd);
    if (!frogData) {
        return {
            header: "Frog",
            displayName: "frog-" + pcd.id.substring(0, 4)
        };
    }
    return {
        displayName: `#${String(frogData.frogId).padStart(3, "00")} ${frogData.name}`
    };
}
exports.getDisplayOptions = getDisplayOptions;
/**
 * The PCD package of the EdDSA Ticket PCD. It exports an object containing
 * the code necessary to operate on this PCD data.
 */
exports.EdDSAFrogPCDPackage = {
    name: EdDSAFrogPCD_1.EdDSAFrogPCDTypeName,
    getDisplayOptions,
    prove,
    verify,
    serialize,
    deserialize
};
