// @pcd/passport-interface/dist/cjs/src/ZupassDefaultSubscriptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.zupassDefaultSubscriptions = void 0;
const pcd_collection_1 = require("@pcd/pcd-collection");
const SubscriptionManager_1 = require("./SubscriptionManager");
exports.zupassDefaultSubscriptions = {
    [SubscriptionManager_1.ZupassFeedIds.Email]: {
        id: SubscriptionManager_1.ZupassFeedIds.Email,
        name: "Zupass Verified Emails",
        description: "PCDs representing Zupass Server's attestation that you own " +
            "the email address corresponding to your Zupass account. Can be " +
            "used to request PCDs from other PCD feeds.",
        partialArgs: undefined,
        credentialRequest: {
            signatureType: "sempahore-signature-pcd"
        },
        permissions: [
            {
                folder: "Email",
                type: pcd_collection_1.PCDPermissionType.DeleteFolder
            },
            {
                folder: "Email",
                type: pcd_collection_1.PCDPermissionType.ReplaceInFolder
            }
        ]
    },
    [SubscriptionManager_1.ZupassFeedIds.Zuzalu_23]: {
        id: SubscriptionManager_1.ZupassFeedIds.Zuzalu_23,
        name: "Zuzalu Tickets",
        description: "EdDSATicketPCDs representing Zuzalu Tickets.",
        partialArgs: undefined,
        credentialRequest: {
            signatureType: "sempahore-signature-pcd"
        },
        permissions: [
            {
                folder: "Zuzalu '23",
                type: pcd_collection_1.PCDPermissionType.DeleteFolder
            },
            {
                folder: "Zuzalu '23",
                type: pcd_collection_1.PCDPermissionType.ReplaceInFolder
            }
        ]
    },
    [SubscriptionManager_1.ZupassFeedIds.Zuconnect_23]: {
        id: SubscriptionManager_1.ZupassFeedIds.Zuconnect_23,
        name: "Zuconnect Tickets",
        description: "EdDSATicketPCDs representing Zuconnect Tickets.",
        partialArgs: undefined,
        credentialRequest: {
            signatureType: "sempahore-signature-pcd"
        },
        permissions: [
            {
                folder: "Zuconnect",
                type: pcd_collection_1.PCDPermissionType.DeleteFolder
            },
            {
                folder: "ZuConnect",
                type: pcd_collection_1.PCDPermissionType.DeleteFolder
            },
            {
                folder: "ZuConnect",
                type: pcd_collection_1.PCDPermissionType.ReplaceInFolder
            }
        ]
    }
};
